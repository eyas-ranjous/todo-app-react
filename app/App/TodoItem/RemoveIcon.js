import React from 'react';
import { useDispatch } from 'react-redux';
import classNames from 'classnames';
import styles from './styles.css';

const RemoveIcon = ({ todo }) => {
  const dispatch = useDispatch();
  const icon = 'âœ˜';

  const removeTodo = () => {
    dispatch({
      type: 'REMOVE_TODO',
      todoId: todo.id
    });
  };

  const classes = classNames(
    styles['todo-item-option'],
    styles['todo-item-remove']
  );

  return (
    <span className={classes} onClick={removeTodo}>
      {icon}
    </span>
  );
};

export default RemoveIcon;
