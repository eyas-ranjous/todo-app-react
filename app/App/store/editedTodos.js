const editedTodos = (state = new Set(), action) => {
  if (action.type === 'EDIT_TODO') {
    return new Set(state).add(action.todoId);
  }

  if (action.type === 'REMOVE_EDITED_TODO') {
    const newState = new Set(state);
    newState.delete(action.todoId);
    return newState;
  }

  return state;
};

export default editedTodos;
