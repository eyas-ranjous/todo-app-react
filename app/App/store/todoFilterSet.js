const todoFilterSet = (state = new Set(), action) => {
  if (action.type === 'UPDATE_STATUS_FILTER_SET') {
    return new Set(action.status);
  }

  return state;
};

export default todoFilterSet;
