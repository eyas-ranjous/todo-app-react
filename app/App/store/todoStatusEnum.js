export const TodoStatusEnum = {
  TODO: 1,
  DONE: 2
};

export const TodoStatus = (state = TodoStatusEnum) => state;
