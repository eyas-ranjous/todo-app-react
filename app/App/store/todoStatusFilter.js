const todoStatusFilter = (state = new Set(), action) => {
  if (action.type === 'SET_TODO_STATUS_FILTER') {
    return new Set(action.status);
  }

  return state;
};

export default todoStatusFilter;
