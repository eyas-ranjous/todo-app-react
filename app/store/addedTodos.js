import { TodoStatusEnum } from './todoStatusEnum';

let nextTodoId = 1;

const addedTodos = (state = new Map(), action) => {
  if (action.type === 'ADD_TODO') {
    const { todo } = action;
    const id = `todo-${nextTodoId}`;
    const status = TodoStatusEnum.TODO;
    nextTodoId += 1;
    return new Map(state).set(id, { ...todo, ...{ id, status } });
  }

  if (action.type === 'UPDATE_TODO') {
    const { todo } = action;

    if (!state.has(todo.id)) return state;

    const updatedTodo = { ...state.get(todo.id), ...todo };
    return new Map(state).set(todo.id, updatedTodo);
  }

  if (action.type === 'REMOVE_TODO') {
    const newState = new Map(state);
    newState.delete(action.todoId);
    return newState;
  }

  if (action.type === 'TOGGLE_TODO_STATUS') {
    if (!state.has(action.todoId)) return state;

    const todo = state.get(action.todoId);
    const { DONE, TODO } = TodoStatusEnum;
    const newStatus = todo.status === TODO ? DONE : TODO;
    const updatedTodo = { ...todo, ...{ status: newStatus } };

    return new Map(state).set(todo.id, updatedTodo);
  }

  return state;
};

export default addedTodos;
