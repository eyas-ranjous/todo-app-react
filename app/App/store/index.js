import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatus';
import todoStatusFilter from './todoStatusFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  todoStatusFilter
});

export default createStore(reducers);
