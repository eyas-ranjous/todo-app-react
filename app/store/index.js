import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatusEnum';
import { TodoFilters } from './todoFiltersEnum';
import todoFilterSet from './todoFilterSet';
import selectedFilter from './selectedFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  TodoFilters,
  todoFilterSet,
  selectedFilter
});

export default createStore(reducers);
