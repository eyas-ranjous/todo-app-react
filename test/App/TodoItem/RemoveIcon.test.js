import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import createStore from '../../../app/store';
import { TodoStatusEnum } from '../../../app/store/TodoStatusEnum';
import RemoveIcon from '../../../app/App/TodoItem/RemoveIcon';

describe('RemoveIcon', () => {
  const store = createStore();
  const todo = {
    id: 'todo-1',
    text: 'test task',
    status: TodoStatusEnum.TODO
  };

  it('renders successfully', () => {
    const wrapper = mount(
      <Provider store={store}>
        <RemoveIcon todo={todo} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('remove todo when icon clicked', () => {
    store.dispatch({ type: 'ADD_TODO', todo: { text: 'test task' } });

    const wrapper = mount(
      <Provider store={store}>
        <RemoveIcon todo={todo} />
      </Provider>
    );
    expect(store.getState().addedTodos.size).toEqual(1);
    wrapper.find('.todo-item-remove').simulate('click');
    expect(store.getState().addedTodos.size).toEqual(0);
  });
});
