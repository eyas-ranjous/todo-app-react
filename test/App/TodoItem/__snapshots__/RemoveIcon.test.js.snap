// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveIcon renders successfully 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <RemoveIcon
    todo={
      Object {
        "id": "todo-1",
        "status": 1,
        "text": "test task",
      }
    }
  >
    <span
      className="todo-item-option todo-item-remove"
      onClick={[Function]}
    >
      âœ˜
    </span>
  </RemoveIcon>
</Provider>
`;
